### ISSUE TITLE
False-positive for W0706 in the case of non-trivial exception inheritance

### ISSUE THREAD

**Issue by nicktimko** [PyCQA/pylint#2484]
If you want to trap a parent exception, but skip certain child exceptions, PyLint sometimes emits "W0706: The except handler raises immediately (try-except-raise)" warnings. In the simplest case where the code raises a direct child but traps the parent, PyLint is happy. Complicating that (trap multiple children, more than one level of inheritance) seems to make things go awry.

Maybe I should take a stab at fixing this; I'm assuming the framework is there because the inheritance is checked on one level, just need to add a few more cases...

### Steps to reproduce
```
class ParentException(Exception):
    pass

class ChildException(ParentException):
    pass

class SiblingException(ParentException):
    pass

class GrandchildException(ChildException):
    pass

def two_levels():
    try:
        pass
    except GrandchildException:
        raise  # W0706: The except handler raises immediately
    except ParentException:
        pass

def no_parent_all_children():
    try:
        pass
    except (ChildException, SiblingException):
        raise  # W0706
    except ParentException:
        pass

def explicit_multilevel():
    """A bit contrived"""
    try:
        pass
    except GrandchildException:
        raise  # W0706
    except (ChildException, ParentException):
        pass

def just_the_leaves():
    """Maybe more practical"""
    try:
        pass
    except (GrandchildException, SiblingException):
        raise  # W0706
    except (ChildException, ParentException):
        pass

def leaves_w_implicit_child():
    try:
        pass
    except (GrandchildException, SiblingException):
        raise  # W0706
    except ParentException:
        pass

def lintsok():
    try:
        pass
    except ChildException:
        raise
    except ParentException:
        pass

def lintsok2():
    try:
        pass
    except GrandchildException:
        raise
    except ChildException:
        pass
```

### Current behavior
```
$ pylint immediateraise.py                    
************* Module immediateraise                                                      
immediateraise.py:16:4: W0706: The except handler raises immediately (try-except-raise)  
immediateraise.py:24:4: W0706: The except handler raises immediately (try-except-raise)  
immediateraise.py:33:4: W0706: The except handler raises immediately (try-except-raise)  
immediateraise.py:42:4: W0706: The except handler raises immediately (try-except-raise)  
immediateraise.py:50:4: W0706: The except handler raises immediately (try-except-raise)  

------------------------------------------------------------------ 
Your code has been rated at 9.12/10 (previous run: 9.12/10, +0.00) 
```

### Expected behavior
No warnings.

### pylint --version output
```
pylint 2.1.1                                  
astroid 2.0.4                                 
Python 3.6.4 (default, Mar 26 2018, 16:21:07) 
[GCC 5.4.0 20160609]                          
```

**Comment by sushobhit27** [comment]
I ll take a look at it soon.

**Comment by sushobhit27** [comment]
@nicktimko you can give it a shot, I might not be able to check this anytime soon.
Below are some pointers to start with.

https://github.com/PyCQA/pylint/blob/faaaafe651699fa333cd5478745136b19b2dbeee/pylint/checkers/exceptions.py#L355
https://github.com/PyCQA/pylint/blob/faaaafe651699fa333cd5478745136b19b2dbeee/pylint/checkers/utils.py#L1043

**Comment by sushobhit27** [comment]
closed via https://github.com/PyCQA/pylint/pull/2508
